app/components/users_modal/users_modal.tsx

```diff
@@ -0,0 +1,234 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React, {MutableRefObject, useCallback, useEffect, useMemo, useReducer, useRef, useState} from 'react';
+import {MessageDescriptor, useIntl} from 'react-intl';
+import {Keyboard, Platform, StyleSheet, View} from 'react-native';
+
+import CompassIcon from '@components/compass_icon';
+import Loading from '@components/loading';
+import SelectedUsersPanel from '@components/selected_users_panel';
+import UserList from '@components/user_list';
+import {useTheme} from '@context/theme';
+import {debounce} from '@helpers/api/general';
+import useNavButtonPressed from '@hooks/navigation_button_pressed';
+import {dismissModal, setButtons} from '@screens/navigation';
+import {filterProfilesMatchingTerm} from '@utils/user';
+
+const style = StyleSheet.create({
+    container: {
+        flex: 1,
+    },
+    searchBar: {
+        marginLeft: 12,
+        marginRight: Platform.select({ios: 4, default: 12}),
+        marginVertical: 12,
+    },
+});
+
+const ACTION_BUTTON = 'action-button';
+const CLOSE_BUTTON = 'close-dms';
+
+const close = () => {
+    Keyboard.dismiss();
+    dismissModal();
+};
+
+type getProfilesError = {
+    users?: undefined;
+    error: unknown;
+}
+
+type getProfilesSuccess = {
+    users: UserProfile[];
+    error?: undefined;
+}
+
+type Props = {
+    buttonText: MessageDescriptor;
+    componentId: string;
+    currentUserId: string;
+    getProfiles: () => Promise<getProfilesSuccess | getProfilesError>;
+    handleRemoveProfile: (id: string) => void;
+    handleSelectProfile: (user: UserProfile) => void;
+    loading: boolean;
+    maxSelectedUsers: number;
+    onButtonTap: (selectedId?: {[id: string]: boolean}) => Promise<boolean>;
+    page: MutableRefObject<number>;
+    searchResults: UserProfile[];
+    selectedIds: {[id: string]: UserProfile};
+    setLoading: (loading: boolean) => void;
+    teammateNameDisplay: string;
+    term: string;
+    tutorialWatched: boolean;
+}
+
+function reduceProfiles(state: UserProfile[], action: {type: 'add'; values?: UserProfile[]}) {
+    if (action.type === 'add' && action.values?.length) {
+        return [...state, ...action.values];
+    }
+    return state;
+}
```

Comment:
Unless you are planning to use in the future more actions, this can be solved with the `function version` of the setState. Here a small example:

```
const [myList, setMyList] = useState([]);
setMyList((prev) => {
    if (someOtherList.length) {
         return [...prev, ...someOtherList];
    }
    return prev;
});
```